autolab/Autolab
# Import course and refactor assessment import

## Summary
This PR implements the import course functionality as described in the export/import course [design doc](https://docs.google.com/document/d/1OExB60sgeOcTN50JQNf9aHsXYZ1iihRPRfYSObNha0o/edit)

## Description

### _Frontend_ ###
* `views/courses/_import_new_course.html.erb` partial for create course tab to upload tar file
* `views/assessments/courseOnboardInstallAsmt.html.erb` displays install assessment table, used in install assessment and import assessment flows
  * `app/views/courses/_assessmentTable.html.erb` partial for assessment table itself
### _Backend_ ###
  * Courses controller:
    * `create_from_tar` endpoint to handle import and create course
      * validates tar file structure (same as assessment import where it checks for a single root directory with a matching .yml file
  * Assessments controller:
    * refactor `importAssessment` to allow for better installation flow with helper function `get_unimported_asmts_from_dir`
    * `importAssessments`  
      * `importAssessmentFromFileSystem` helper function to import + save multiple assessments at once

New course flow:

- New Course Page has two tabs, one for creating from scratch, the other for importing via tar:
  
![Screenshot 2023-12-03 at 11 46 09 AM](https://github.com/autolab/Autolab/assets/25730111/4f431658-0a68-4ec3-a5c0-4946b6c50ae1)
![Screenshot 2023-12-03 at 11 46 12 AM](https://github.com/autolab/Autolab/assets/25730111/5c767722-6e1f-415f-8939-40e1dcf26260)

- Upon importing a course, the user will then be redirected to a page where they can import assessments extracted from the tar
![Screenshot 2023-12-03 at 11 46 24 AM](https://github.com/autolab/Autolab/assets/25730111/fc90d10f-24a6-4834-9976-74357e1cc6a8)
![Screenshot 2023-12-03 at 11 46 34 AM](https://github.com/autolab/Autolab/assets/25730111/8f66b958-1096-459a-8321-6ac1fb79e24c)

- After importing assessments, they can edit the course configuration using the existing "edit" page

![Screenshot 2023-12-03 at 11 46 44 AM](https://github.com/autolab/Autolab/assets/25730111/804d780e-1bd8-4b78-8409-5cffc2cdd80f)

Video of flow:

https://github.com/autolab/Autolab/assets/25730111/29a8c19e-b36f-485a-952c-b7c551534796

Assessment Import Refactor:
- code to get unimported assessments in file system moved from `install_assessment` to`get_unimported_asmts_from_dir`
- New controller action `importAssessments` to handle importing multiple assessments and return the statuses of each (to handle being able to display errors from each assessment import on the frontend)
- `importAssessment` refactored to `importAssessmentFromFileSystem` to import multiple assessments from file system (which both tar and directory based import use), return an array of results
  - (also will add info comment if for example a config rb file was automatically generated)
- Importing assessments (via tar and file system) now also use the new `importAssessmentFromFileSystem`
- Importing via tar has same behavior as before with redirect via `handleImportResults` action which takes message from `importAssessmentFromFileSystem` and redirects, maintaining previous behavior 

Import Success:
![Screenshot 2023-12-03 at 11 46 34 AM](https://github.com/autolab/Autolab/assets/25730111/ad68df76-b637-42de-9cab-dcedcbe361dc)

Import Failure:
![Screenshot 2023-12-03 at 1 01 08 PM](https://github.com/autolab/Autolab/assets/25730111/c1da7b80-b22b-4f3d-a423-3a295ad17a24)


Import Success with conditions:
![Screenshot 2023-12-03 at 1 02 49 PM](https://github.com/autolab/Autolab/assets/25730111/0f12bc1e-8bb2-4c91-96c9-a7b65c077fb7)


## Motivation and Context


Part of the effort to add functionality of exporting and importing courses for the transition to Autolab 3.0.

## How Has This Been Tested?



### _Manual Testing_ ###
Get a course tar
* Create a tarball from scratch
  * Create a directory named as the course name
  * Add 'course-name.yml' with the relevant configurations
* Export existing course
  * Checkout `features/import-export-course-fa23` branch, go to a course and export it in manage course
  * Delete the exported course
* Using existing tar file:
  * use [pog-s21_20231026.tar.zip](https://github.com/autolab/Autolab/files/13539584/pog-s21_20231026.tar.zip)
  * (has various assessments)

Import the tar file
* Go to 'Create new course' in Manage Autolab dropdown
  * Go to 'Import Course' tab
  * Upload the tar file and input an instructor email
  * Press 'Create'
  * see that course attributes match as expected, and assessments are put into the course directory to be imported
  * See that on import of multiple assessments, error messages will display for bad imports and will display success for good imports
  * verify that new courses can still be created as before

### _Automated Tests_ ###
* Functionality tests for `create_from_tar` endpoint in `spec/controllers/course_controller_spec.rb`
  * verifies a course is created
  * verifies that various errors are handled 

## Types of changes


## Checklist:
